[@page] `size` should only parse as a descriptor, not a global CSS property

nt1m opened the pull request:
#### 4d0b0c1943c5a4e05b60b4dfe1f09c9e310eca54
<pre>
[@page] `size` should only parse as a descriptor, not a global CSS property
<a href="https://bugs.webkit.org/show_bug.cgi?id=239965">https://bugs.webkit.org/show_bug.cgi?id=239965</a>
<a href="https://rdar.apple.com/92963022">rdar://92963022</a>

Reviewed by Darin Adler.

Stop parsing `size` as a CSS property in the global context, instead consider it as a descriptor for `@page`.

This creates new parsing path for parsing declarations for StyleRuleType::Page, and uses that instead of the normal StyleRuleType::Style path

Now that this registers `size` as a descriptor in CSSProperties.json instead of global property, we have to move the code from StyleBuilderCustom
to StyleBuilder, given CSSProperties.json won&apos;t generate builder code for descriptors.

Also do a couple of drive-by cleanups.

* LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-CSSStyleDeclaration-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-page/parsing/size-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-page/parsing/size-invalid.html: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeKeyframesRule):
(WebCore::CSSParserImpl::consumePageRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::CSSParserImpl::consumeDeclaration):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue):
(WebCore::CSSPropertyParser::parsePageDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
(WebCore::Style::pageSizeFromName):
(WebCore::Style::Builder::applyPageSizeDescriptor):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSize):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::mmLength): Deleted.
(WebCore::Style::BuilderCustom::inchLength): Deleted.
(WebCore::Style::BuilderCustom::getPageSizeFromName): Deleted.
(WebCore::Style::BuilderCustom::applyInheritSize): Deleted.
(WebCore::Style::BuilderCustom::applyInitialSize): Deleted.
(WebCore::Style::BuilderCustom::applyValueSize): Deleted.

Canonical link: <a href="https://commits.webkit.org/281720@main">https://commits.webkit.org/281720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28eb3836739ef01d77b83ad4479efe9245ea042e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60771 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40130 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64702 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11318 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62901 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47806 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11593 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49145 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7859 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62805 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37375 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/52651 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29977 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34062 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9877 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10231 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55901 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10175 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66431 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4715 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10008 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56513 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52619 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56695 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13552 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3912 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35935 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37017 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38110 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -9969,6 +9952,23 @@
                 }
             }
+        },
+        "@view-transition": {
```
Why is this in this pull request? Seems not part of the `size` change?

nt1m replied:
I did an alpha sort

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3961,6 +3961,10 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
     case CSSPropertyPaddingLeft:
         return zoomAdjustedPaddingOrMarginPixelValue<&RenderStyle::paddingLeft, &RenderBoxModelObject::computedCSSPaddingLeft>(style, renderer);
+    case CSSPropertyPage:
+        // FIXME: The page CSS property was never not fully implemented.
```
What does “never not fully implemented” mean? But whatever it means, do we really need this mentioned in a comment here? It’s enough to know the computed style needs to be implemented. Maybe there’s a point to be made but I don’t understand what it is.

nt1m replied:
We don't actually do anything with the page names atm, but maybe that's not super valuable info indeed in this context

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3961,6 +3961,10 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
         return zoomAdjustedPaddingOrMarginPixelValue<&RenderStyle::paddingLeft, &RenderBoxModelObject::computedCSSPaddingLeft>(style, renderer);
+    case CSSPropertyPage:
+        // FIXME: The page CSS property was never not fully implemented.
+        // The computed style should be fixed, since the property works everywhere.
```
I don’t understand what “works everywhere” means here. What’s the point of this comment? Maybe the comment makes sense in the change log, because the old code erroneously said `for @page`, but in the code itself I don’t know what good it does.

nt1m replied:
`page` is a CSS property, not an at-rule descriptor, the old comment was wrong by suggesting it was specific to @page

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -212,18 +212,20 @@ bool CSSPropertyParser::parseValue(CSSPropertyID propertyID, bool important, con
 
     bool parseSuccess;
-    if (ruleType == StyleRuleType::FontFace)
+    if (ruleType == StyleRuleType::CounterStyle)
```
Should we change this to a `switch` at some point?

nt1m replied:
Good idea

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -630,6 +624,23 @@ bool CSSPropertyParser::parseFontPaletteValuesDescriptor(CSSPropertyID property)
+
+    RefPtr parsedValue = CSSPropertyParsing::parsePageDescriptor(m_range, property, m_context);
+    if (!parsedValue)
+        return parseValueStart(property, important);
```
Does this include the `atEnd` check?

nt1m replied:
Other similar parsing functions don't (presumably because there's one at a level high-up), but I can check

nt1m replied:
parseValueStart does include its own end checks

darinadler commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -557,6 +563,146 @@ RefPtr<CSSCustomPropertyValue> Builder::resolveCustomPropertyValue(CSSCustomProp
+
+void Builder::applyPageSizeDescriptor(CSSValue& value)
+{
+    m_state.style().resetPageSizeType();
```
I kind of wish for a better idiom where we have a function that parses into a `std::optional<PageSizeDdscriptor>`, which holds the type and the lengths in a `struct`, and then write the values into `m_state.style()` afterward based on the return value. But maybe that’s not the style of the `Builder` code.

nt1m replied:
@page is fairly old code, so it's using quite unconventional patterns. It would be a larger project to move it away from RenderStyle.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-CSSStyleDeclaration-expected.txt:
```diff
@@ -1323,7 +1323,7 @@ PASS shape-outside: _camel_cased_attribute v. CSS.supports
 PASS shape-rendering: _camel_cased_attribute v. CSS.supports
 PASS shape-rendering: _dashed_attribute v. CSS.supports
-PASS size: _camel_cased_attribute v. CSS.supports
+FAIL size: _camel_cased_attribute v. CSS.supports assert_equals: expected true but got false
```
I checked, and other descriptors fail in the same way.

nt1m committed:
[@page] `size` should only parse as a descriptor, not a global CSS property
https://bugs.webkit.org/show_bug.cgi?id=239965
rdar://92963022

Reviewed by Darin Adler.

Stop parsing `size` as a CSS property in the global context, instead consider it as a descriptor for `@page`.

This creates new parsing path for parsing declarations for StyleRuleType::Page, and uses that instead of the normal StyleRuleType::Style path

Now that this registers `size` as a descriptor in CSSProperties.json instead of global property, we have to move the code from StyleBuilderCustom
to StyleBuilder, given CSSProperties.json won't generate builder code for descriptors.

Also do a couple of drive-by cleanups.

* LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-CSSStyleDeclaration-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-page/parsing/size-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-page/parsing/size-invalid.html: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeKeyframesRule):
(WebCore::CSSParserImpl::consumePageRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::CSSParserImpl::consumeDeclaration):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue):
(WebCore::CSSPropertyParser::parsePageDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
(WebCore::Style::pageSizeFromName):
(WebCore::Style::Builder::applyPageSizeDescriptor):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSize):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::mmLength): Deleted.
(WebCore::Style::BuilderCustom::inchLength): Deleted.
(WebCore::Style::BuilderCustom::getPageSizeFromName): Deleted.
(WebCore::Style::BuilderCustom::applyInheritSize): Deleted.
(WebCore::Style::BuilderCustom::applyInitialSize): Deleted.
(WebCore::Style::BuilderCustom::applyValueSize): Deleted.

Canonical link: https://commits.webkit.org/281720@main
